### DISCUSSION TITLE
Go: FunctionNode abstract methods

### DISCUSSION THREAD

**Discussion by jason-invision** [github/securitylab#188] [General] [answered]
Extending the FunctionNode type is a bit complex due to the number of abstract methods that must be implemented.

```
class MyFunctionNode extends DataFlow::FunctionNode {
    MyFunctionNode() {
        this.getParameter(0).getType().hasQualifiedName("pkg", "type")
        and this.getParameter(1).getType().(PointerType).getBaseType().hasQualifiedName("pkg", "type")
    }

    override string getName() { result = getName() }
  
    override DataFlow::ReceiverNode getReceiver() { result = getReceiver() }
  
    override string toString() { result = "MyFunctionNode" }
  
    override DataFlow::ResultNode getAResult() { result = getAResult() }

    override DataFlow::ParameterNode getParameter(int i) { result = getParameter(i) }
```

Is there a way to create a more generalized implementation of these methods in order to reduce the number of methods that must be implemented?

**Comment by max-schaefer** [comment]
Hm, yes, that's a little cumbersome. Note that you can just override all of these member predicates as `none()`; the implementations in the sibling classes will do the actual work.

I think we may want to switch that class to using the `::Range` pattern. What do you think, @smowton?

**Comment by smowton** [comment, marked as answer]
Seems like a good idea: https://github.com/github/codeql-go/pull/362

**Reply by smowton** [reply]
This is now merged
